### PR TITLE
nss: do not set errno to ENOENT or ESRCH when NSS_STATUS_NOTFOUND

### LINKED ISSUE #9585: nss-systemd's getpwent sets errno to ENOENT
**systemd version the issue has been seen with**
> v209

**Used distribution**
 > Arch Linux (package systemd 239.0-2)

**Expected behaviour you didn't see**
When using a `getpwent()` loop in a C program, I expect it to end without setting errno if the enumeration was successful. Here is a C program which illustrates such a loop:
```c
#include <sys/types.h>
#include <pwd.h>
#include <errno.h>
#include <stdio.h>

int main(void)
{
    struct passwd *pw;
    errno = 0;
    setpwent();
    while ((pw = getpwent()) != NULL) {
        printf("%d: %s\n", pw->pw_uid, pw->pw_name);
    }
    printf("getpwent errno = %d (%m)\n", errno);
    endpwent();
    return 0;
}
```

**Unexpected behaviour you saw**
The C program shows "getpwent errno = 2 No such file or directory" (`errno` 2 is `ENOENT`). This suggests an error happened, even though everything else seems to work fine.

If nss-systemd is removed from `/etc/nsswitch.conf`, `errno` remains 0.

**Steps to reproduce the problem**
* Configure `/etc/nsswitch.conf` to use `nss-systemd` for `passwd` ("`passwd: files mymachines systemd`")
* Write the C program into a file.
* Compile it using `gcc` for example.
* Run it and watch for the resulting "`getpwent errno = ...`" line.

**Additional information**
`errno` is set to `ENOENT` in `_nss_systemd_getpwent_r` when there are no more entries to enumerate: https://github.com/systemd/systemd/blob/de7436b02badc82200dc127ff190b8155769b8e7/src/nss-systemd/nss-systemd.c#L780-L784

This issue has also been reported to glibc's bug tracker (https://sourceware.org/bugzilla/show_bug.cgi?id=23410). It has been categorized as invalid because:
> NSS_STATUS_NOTFOUND is not an error condition and the nss module must not modify errno.

Therefore it seems that statement `*errnop = ENOENT;` needs to be removed from `_nss_systemd_getpwent_r` and `_nss_systemd_getgrent_r`. What is your opinion on this matter?

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #9585, and https://sourceware.org/bugzilla/show_bug.cgi?id=23410.

### CODE REVIEW

**Comment by yuwata** [comment]
> NSS_STATUS_NOTFOUND is not an error condition and the nss module must not modify errno.

How about `NSS_STATUS_SUCCESS`? We should neither set 0 to `errno` in that case?

**Comment by filbranden** [comment]
> How about `NSS_STATUS_SUCCESS`? We should neither set 0 to `errno` in that case?

Correct. (As we're not setting it currently.)

I think typically it's up to the caller (in this case glibc) to reset `errno` to 0 before calling into modules.

It's possible they're also keeping a single `errno` for multiple NSS calls so they might want to see the "last" error that happened.

I'm almost positive that just not setting it (except for ERANGE above) is the correct approach.

**Comment by poettering** [comment]
Uh, this change is pretty explicitly *against* glibc documentation:

See the discussion here:

https://www.gnu.org/software/libc/manual/html_node/NSS-Modules-Interface.html

It explicitly lists ENOENT as a possible error code for NSS_STATUS_NOTFOUND. It appears the glibc folks are contradicting the very explicit docs here...

**Comment by poettering** [comment]
That said all our other codes do set errno to 0 in case of NOTFOUND, for example `_nss_systemd_getpwnam_r()`, hence I figure regardless whether the glibc docs and devs contradict each other I figure it's a good idea to merge this.

**Comment by poettering** [comment]
But then again, many of our other NSS calls do set ENOENT in case of NOTFOUND...

for example nss-myhostname... If we change anything here we should probably bring this all in sync, across all our four NSS modules...

**Comment by poettering** [comment]
lgtm failures appear to be network issues.

**Comment by poettering** [comment]
rawhide CI failure appears to be caused by statx() glibc issue, i.e. is unrelated

**Comment by yuwata** [comment]
Updated. Now all the cases where `NSS_STATUS_NOTFOUND` are returned set `errno` 0.

**Comment by yuwata** [comment]
@filbranden The reason I still set errno 0 is, the document @poettering tells saids
```
NSS_STATUS_NOTFOUND | SUCCESS | There are no entries. Use this to avoid returning errors for inactive services which may be enabled at a later time. This is not the same as the service being temporarily unavailable.
```
I think 'SUCCESS' means assigning 0, no?

**Comment by codonell** [comment]
Keep *errnop untouched. You should not be setting it to zero. In the manual the entry for "NSS_STATUS_NOTFOUND + SUCCESS" means "Return the code NSS_STATUS_NOTFOUND and explicitly do not set *errnop", this is specifically called out for inactive services where a server may not yet be connected to a data source, and this is not an error.

**Comment by filbranden** [comment]
Considering the line just adjacent to that one in that document was incorrect, I wouldn't be surprised if that line was too...

**Comment by codonell** [comment]
@filbranden @poettering Please note that there is no contradiction or error in the manual pages. The references there are for the NSS subsystem as a whole and does not cover the specific functions. We can and should expand our documentation to include more detailed functional definitions like this, but for now we don't have such detailed documentation.

**Comment by yuwata** [comment]
I've force-pushed a revised version.

@codonell Thank you for the comment. Can I ask one more question? How about `*h_errnop` when `NSS_STATUS_SUCCESS`? Currently, we set `NETDB_SUCCESS`.

**Comment by yuwata** [comment]
Updated. Now also `h_errno` is not touched when `NSS_STATUS_SUCCESS` is returned.

**Comment by poettering** [comment]
There's also this:

https://bugzilla.redhat.com/show_bug.cgi?id=1125975 and e70df46b9721a3d025e7a0b4ffb5893cbde5e55d

This was fixed in glibc in 2.27 afaics, which has been closed only beginning of this year. Hence I'd rather not take the risk on that...

**Comment by codonell** [comment]
@yuwata @poettering I agree. Until the 2.27 fix is more widely propagated you want to set h_errno and *h_errnop, only because of the getaddrinfo defects. If you run into any more problems in this space please open a bug and we'll handle it.

File: src/nss-resolve/nss-resolve.c [hunk 1]
```diff
@@ -258,6 +250,10 @@ enum nss_status _nss_resolve_gethostbyname4_r(
         *errnop = -r;
```
Comment (filbranden) [inline src/nss-resolve/nss-resolve.c:251]: A few lines up:

```c
         /* Explicitly reset all error variables */
         *errnop = 0;
         *h_errnop = NETDB_SUCCESS;
         h_errno = 0;
```

I thought even though we're keeping the reset of h_errno and *h_errnop, we're no longer setting *errnop = 0 anywhere... right?

File: src/nss-resolve/nss-resolve.c [hunk 2]
```diff
@@ -427,23 +415,27 @@ enum nss_status _nss_resolve_gethostbyname3_r(
...
-        h_errno = 0;
-
         if (ttlp)
                 *ttlp = 0;
 
         if (canonp)
                 *canonp = r_name;
 
+        /* Explicitly reset all error variables */
+        *errnop = 0;
```
Comment (filbranden) [inline src/nss-resolve/nss-resolve.c]: Same as above... Remove this last *errnop = 0?

File: src/nss-resolve/nss-resolve.c [hunk 3]
```diff
@@ -623,6 +607,10 @@ enum nss_status _nss_resolve_gethostbyaddr2_r(
         *errnop = -r;
```
Comment (filbranden) [inline src/nss-resolve/nss-resolve.c:610]: And a few lines above here too...

**Comment by yuwata** [comment]
@filbranden I am not sure but I guess from the code
https://github.com/bminor/glibc/blob/53250a21b81474ef4e78090a4a9a63d8471e1091/sysdeps/posix/getaddrinfo.c#L262
that `errno` needs to be reset. Let's wait comments from @poettering and/or @codonell.

File: src/nss-resolve/nss-resolve.c (hunk 1 above)
Comment (codonell) [reply src/nss-resolve/nss-resolve.c:251]: Drop the *errnop reset. You should not be setting errno to 0. Keep the h_errnop/h_errno reset for rhbz 1125975.

File: src/nss-resolve/nss-resolve.c (hunk 2 above)
Comment (codonell) [reply src/nss-resolve/nss-resolve.c]: Don't reset *errnop to zero.

**Comment by codonell** [comment]
@poettering @yuwata Commented. I re-reviewed the upstream glibc 2.27 fixes, and they only impact h_errno and *h_errnop. In general one must not set *errnop or errno to zero, this is not allowed by the runtime, you should only set it to non-zero in the event of an error. In the case of *h_errnop and h_errno, you can reset these to work around rhbz#1125975. That should cover your cases.

**Comment by poettering** [comment]
@codonell did you see the link into the glibc sources @yuwata linked:

https://github.com/bminor/glibc/blob/53250a21b81474ef4e78090a4a9a63d8471e1091/sysdeps/posix/getaddrinfo.c#L262

It's the macro that glibc uses to build getaddrinfo() and friends from. It checks errno == 0 in various places. This suggests that the function call our nss modules support either need to reset errno to zero explicitly before retruning or at least save errno when beginning and restoring it before returning. 

What's the story there? You are saying we shall not reset errno to zero, but should we make sure that before and after our calls it remains unmodified?

**Comment by codonell** [comment]
@poettering @yuwata This is glibc bug [swbz#21915](https://sourceware.org/bugzilla/show_bug.cgi?id=21915) fixed in 2.27, 2.26, and 2.25. However, you may still not set errno to 0, but you may, as a workaround for the bug, save and restore errno if you call routines whose failure may in turn set errno on failure, and you do not intend that value of errno to propagate to the caller as the intended failure (a clobbered errno). The DL_CALL_FCT() wrapper would not call anything that sets errno either (wraps function call for mcount profiling purposes). In the end setting errno to 0 can hide other serious problems which we should not hide, like other functions failing, and you prevent that from propagating correctly when the bug is fixed.

**Comment by poettering** [comment]
@codonell excellent! thank you for getting back to us.

@yuwata following @codonell's suggestions i figure we should stop touching errno directly, but we should instead add out PROTECT_ERRNO macro everywhere so that all our NSS functions will return with the same errno value as was set when they were entered.

**Comment by yuwata** [comment]
@codonell Thank you for the comments.

@poettering OK. I will update this in that way.

**Comment by yuwata** [comment]
Updated. Now `PROTECT_ERRNO` is added to all nss module functions.

**Comment by poettering** [comment]
Excellent! Thanks for reworking this!

**Pull request merged**